File: retrocpu_boot_monitor/test/mn1613/handshake/handshake_main_test.ts
/**
 * g_handshake_interrupt_handler（IO→CPU コマンド 10h–18h）
 * 根拠: HandShake.mdc / boot_monitor.mdc / test_framework.mdc
 */
import {
  createSessionFromSettings,
  expect,
  test,
  type IoBoardHandshakeMock,
  type Mn1613AsmSession,
} from "../../../../retrocpu_test_framework_ts/src/index.js";
import {
  mn1613MonHandshakeSettings,
  withMn1613CpuLog,
} from "../mn1613_mon_settings.js";

const BASE_REGS = {
  R2: 0x2222,
  R3: 0x3333,
  R4: 0x4444,
} as const;

const session: Mn1613AsmSession = createSessionFromSettings(
  withMn1613CpuLog(mn1613MonHandshakeSettings, import.meta.url),
);

/**
 * g_main 済み＋ ioMock handshake で 1 ケースを実行する。
 * @param fn 本体（session / mock 利用可）
 */
async function withCase(
  fn: (s: Mn1613AsmSession, mock: IoBoardHandshakeMock) => Promise<void>,
): Promise<void> {
  session.reload();
  try {
    await session.runInit();
    await fn(session, session.requireHandshakeMock());
  } finally {
    await session.detachIoMock();
  }
}

/**
 * IO が HSHK_IN_REQ を上げるまで待つ（受理より先に依頼を出す）。
 * @param mock IO モック
 * @param timeoutMs 上限 ms
 */
async function waitReq1(
  mock: IoBoardHandshakeMock,
  timeoutMs = 2000,
): Promise<void> {
  const t0 = Date.now();
  while (mock.bus.HSHK_IN_REQ !== 1) {
    if (Date.now() - t0 > timeoutMs) {
      throw new Error("timeout waiting HSHK_IN_REQ");
    }
    await new Promise<void>((resolve) => setTimeout(resolve, 0));
  }
}

/**
 * ハンドラを `` で呼び、IO→CPU 交換と並行する。
 * IO が REQ_1 を上げてから CPU を動かす（ENA0 競合を避ける）。
 * @param mock IO モック
 * @param toCpu IO→CPU フレーム
 * @param fromCpu CPU→IO 待ちバイト数
 * @param thenToCpu 応答後の追加 IO→CPU
 * @returns CPU→IO 応答
 */
async function callHandler(
  mock: IoBoardHandshakeMock,
  toCpu: Uint8Array,
  fromCpu: number,
  thenToCpu?: Uint8Array,
): Promise<Uint8Array> {
  const io = mock.exchangeWithCpu(toCpu, fromCpu, thenToCpu);
  await waitReq1(mock);
  await session.call("g_handshake_interrupt_handler", {
    registers: { ...BASE_REGS },
  });
  return io;
}

test("0x0F（<0x10）はディスパッチせず完了する", async () => {
  await withCase(async (s, mock) => {
    const reply = await callHandler(mock, Uint8Array.from([0x0f]), 0);
    expect(reply.length).toBe(0);
    s.expectRegisters({ R0: 0, R4: 0x4444 });
  });
});

test("0x44 未実装コマンドは NG も返さず完了する", async () => {
  await withCase(async (s, mock) => {
    const reply = await callHandler(mock, Uint8Array.from([0x44]), 0);
    expect(reply.length).toBe(0);
    s.expectRegisters({ R0: 0, R4: 0x4444 });
  });
});

test("12h 実行指示は 5B（pad 含む）を読み NG を返す", async () => {
  await withCase(async (s, mock) => {
    const reply = await callHandler(
      mock,
      Uint8Array.from([0x12, 0, 0, 0x02, 0x00, 0x00]),
      1,
    );
    expect(Array.from(reply)).toEqual([0x01]);
    s.expectRegisters({ R0: 0, R4: 0x4444 });
  });
});

test("0x48（廃止の CPU状態取得）は NG も返さず完了する", async () => {
  await withCase(async (s, mock) => {
    const reply = await callHandler(mock, Uint8Array.from([0x48]), 0);
    expect(reply.length).toBe(0);
    s.expectRegisters({ R0: 0, R4: 0x4444 });
  });
});

File: retrocpu_boot_monitor/test/mn1613/handshake/handshake_main_test.rs
use std::sync::Arc;

use retrocpu_test_framework_rs::{
    create_session_from_settings, CallOptions, CallRegisters, CodeTestIoMockEntry, FrameworkError,
    IoBoardHandshakeMock, JsonTestSettings, Mn1613AsmSession,
};

fn base_regs() -> CallRegisters {
    CallRegisters {
        r2: Some(0x2222),
        r3: Some(0x3333),
        r4: Some(0x4444),
        ..Default::default()
    }
}

fn handshake_settings() -> JsonTestSettings {
    let mut s = super::mn1613_rs_settings();
    s.io_mock = Some(vec![CodeTestIoMockEntry::Handshake]);
    s
}

fn with_case<F>(f: F) -> Result<(), FrameworkError>
where
    F: FnOnce(&mut Mn1613AsmSession, Arc<IoBoardHandshakeMock>) -> Result<(), FrameworkError>,
{
    let mut s = create_session_from_settings(&handshake_settings(), None)?;
    s.run_init()?;
    let mock = s.require_handshake_mock()?;
    let ret = f(&mut s, Arc::clone(&mock));
    s.detach_io_mock();
    ret
}

fn call_handler(
    s: &mut Mn1613AsmSession,
    mock: &Arc<IoBoardHandshakeMock>,
    to_cpu: &[u8],
    from_cpu_len: usize,
) -> Result<Vec<u8>, FrameworkError> {
    mock.run_io_handler_exchange(to_cpu, from_cpu_len, || {
        let _ = s.call(
            "g_handshake_interrupt_handler",
            CallOptions {
                registers: Some(base_regs()),
                ..Default::default()
            },
        )?;
        Ok(())
    })
}

#[test]
fn cmd_0x0f_is_ignored_without_reply() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        let reply = call_handler(s, &mock, &[0x0f], 0)?;
        assert!(reply.is_empty());
        s.expect_registers(
            &CallRegisters {
                r0: Some(0),
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}

#[test]
fn cmd_0x44_finishes_without_reply() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        let reply = call_handler(s, &mock, &[0x44], 0)?;
        assert!(reply.is_empty());
        s.expect_registers(
            &CallRegisters {
                r0: Some(0),
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}

#[test]
fn cmd_0x12_drains_payload_and_returns_ng() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        let reply = call_handler(s, &mock, &[0x12, 0x00, 0x00, 0x02, 0x00, 0x00], 1)?;
        assert_eq!(reply, vec![0x01]);
        s.expect_registers(
            &CallRegisters {
                r0: Some(0),
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}

#[test]
fn cmd_0x48_finishes_without_reply() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        let reply = call_handler(s, &mock, &[0x48], 0)?;
        assert!(reply.is_empty());
        s.expect_registers(
            &CallRegisters {
                r0: Some(0),
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}
